Add few more fields to song information

pmphfm opened the pull request:
I haven't tested it on a windows box.
This pr also adds vscode settings which I don't think it is worth to split into its own pr.

Test: Changed different file formats and ran tests run against those files.

agersant commented on docs/swagger/polaris-api.json:
```diff
@@ -1417,6 +1417,26 @@
                     "duration": {
                         "type": "integer",
                         "example": 571
+                    },
```
Thank you for not forgetting to update this!
I do have one minor request, could you keep the example values consistent with each other? The current ones are using a song from [this record](https://www.discogs.com/Stratovarius-Destiny/release/873481). Switching to a different song to have different lyricist vs composer is ok.

agersant commented on src/app/index/metadata.rs:
```diff
@@ -42,6 +47,11 @@ impl From<id3::Tag> for SongTags {
+		let lyricist = tag.get_text("TEXT");
+		let composer = tag.get_text("TCOM");
+		let genre = tag.genre().map(|s| s.to_string());
+		let category = tag.get_custom_field("Category");
```
I don't think `Category` is a standard or widespread field. Unless I am incorrect here, I would prefer not to include it. The example value in the swagger file is `Filmi` which I would expect to find under Genre (similar to game or movie soundtracks).

Sources:
https://wiki.hydrogenaud.io/index.php?title=Tag_Mapping
https://picard-docs.musicbrainz.org/en/appendices/tag_mapping.html

pmphfm replied:
I dropped category. 

(I guess I will have to maintain a patch for category. Maybe game and movie sound reasonable to you as genre but to some of us they are not. Even the genre as defined by id3 standards is not sufficient for genre itself. case in point: id3 has "classical" but India has at least two major classical forms and each having multiple sub-genre. For long I have been stuffing "filmi" in different tag fields. Rust/polaris allowing me(a firmware/embedded guy) to write a music server with ease) so maybe category can be a new 'custom" field.)

agersant replied:
>  Even the genre as defined by id3 standards is not sufficient for genre itself. case in point: id3 has "classical" but India has at least two major classical forms and each having multiple sub-genre.
The ID3 genre list ([this disaster](https://en.wikipedia.org/wiki/ID3#Genre_list_in_ID3v1[12])) is a leftover from ID3v1 which isn't used much these days. I wholeheartedly agree that would be inadequate, even for popular European/American music. In current versions, the `Genre` field is a text frame (`TCON`) that supports multiple values (see https://id3.org/id3v2.4.0-frames)
You can tag a song with both `Karnāṭaka` and `Tanam` in the Genre field, and even more values if necessary. At least that's what I do for genre and subgenres in my collection.

agersant commented on src/app/index/metadata.rs:
```diff
@@ -79,6 +94,46 @@ pub fn read(path: &Path) -> Option<SongTags> {
+		}
+	}
+
+	fn get_custom_field(&self, field_name: &str) -> Option<String> {
```
If we do end up keeping the category field, I would prefer writing this block using `match` instead of `unwrap` calls.

pmphfm replied:
Dropped the entire function.

agersant commented on src/app/index/metadata.rs:
```diff
@@ -147,6 +202,11 @@ fn read_ape(path: &Path) -> Result<SongTags> {
 	let year = tag.item("Year").and_then(read_ape_i32);
 	let disc_number = tag.item("Disc").and_then(read_ape_x_of_y);
 	let track_number = tag.item("Track").and_then(read_ape_x_of_y);
+	let lyricist = tag.item("LYRICIST").and_then(read_ape_string);
```
Assuming the names are not case sensitive, can you keep the case consistent between them?

pmphfm replied:
Surprisingly they are. (I tried kid3 and mp3tag.) Let me know.

agersant commented on src/app/index/metadata.rs:
```diff
@@ -286,6 +382,11 @@ fn reads_file_metadata() {
 		duration: None,
 		year: Some(2016),
 		has_artwork: false,
+		lyricist: Some("TEST LYRICIST".into()),
```
Thank you for updating the tests ❤️

pmphfm replied:
Tests are fun to write :)

agersant reviewed (changes requested):
Left a few comments (mostly minor), thank you for working on this!

pmphfm commented:
Thanks for the comments. I might address them by EOD. (Also there is some learning curve with github workflows)

pmphfm commented:
Note: In addition to addressing comments I renamed the migration directory to today's date.

agersant commented:
Looks like one test is failing:
![image](https://user-images.githubusercontent.com/817256/117087255-bc43d000-ad03-11eb-9e30-5c003e6d554d.png)

Can you update the test either by fixing the test-case or dealing with m4a files? Thanks!

pmphfm commented:
I couldn't reproduce the issue locally. I tried it on two different boxed(both linux) and tried pulling from the merge branch. I am not sure what is going on. Wonder some issue with my github workflow?

Saecki commented:
I pulled your branch and everything was fine. After rebasing on top of master I got, the exact error the github actions show. Github runs the actions after the merging into/rebasing on top of master. The errors occur because because the commits that are new in master but not yet in your branch added `aiff` and `wave` support and changed some of the id3 code.

pmphfm commented:
Thanks for pointing it out. Fixed the test case but I think I messed up something while fetch-rebase cycle. Now 4 commits are showing up in this pull request. This is what I did.

- Fetched from master the latest changes into pull-requests branch
- Fixed the test case. Committed and `rebase -i` to float the commit to the top
- Force pushed the changes.

agersant commented:
Thanks for taking a look! I'm not super good at Git either so I don't have a recommendation on where to go from here (there is a conflict on `test-data/formats/sample.aif`) preventing merge.

Don't worry about the unwanted commits, I will use `Squash And Merge` at the end so all your changes will appear as a single commit in the final history.

Saecki commented:
> Fixed the test case. Committed and rebase -i to float the commit to the top

I think in order to preserve the existing commits you first have to rebase and then apply and commit changes to the branch.
